skipruntime-ts/helpers/src/join.ts

```diff
@@ -0,0 +1,156 @@
+import type {
+  EagerCollection,
+  Json,
+  JsonObject,
+  Values,
+  Mapper,
+  DepSafe,
+} from "@skipruntime/core";
+
+class JoinOneMapper<
+  IdProperty extends keyof VLeft,
+  JoinedProperty extends string,
+  K extends Json,
+  VLeft extends JsonObject & { [P in IdProperty]: Json },
+  VRight extends Json,
+> implements
+    Mapper<
+      K,
+      VLeft,
+      K,
+      Omit<VLeft, IdProperty> & Record<JoinedProperty, VRight>
+    >
+{
+  constructor(
+    private readonly right: EagerCollection<VLeft[IdProperty], VRight>,
+    private readonly on: IdProperty,
+    private readonly name: JoinedProperty,
+  ) {}
+
+  mapEntry(
+    key: K,
+    values: Values<VLeft>,
+  ): Iterable<[K, Omit<VLeft, IdProperty> & Record<JoinedProperty, VRight>]> {
+    return values.toArray().map((v: VLeft) => {
+      const { [this.on]: key_right, ...value_left } = v;
+      const value_right = {
+        [this.name]: this.right.getUnique(key_right),
+      } as Record<JoinedProperty, VRight>;
+      const value_out = { ...value_left, ...value_right } as const;
+      return [key, value_out];
+    });
+  }
+}
+
+export function join_one<
+  IdProperty extends keyof V1,
+  JoinedProperty extends string,
+  K extends Json,
+  V1 extends JsonObject & { [P in IdProperty]: Json },
+  V2 extends Json,
+>(
+  left: EagerCollection<K, V1>,
+  right: EagerCollection<V1[IdProperty], V2>,
+  options: {
```

Comment:
I don't understand why these 2 args are packed into a record. Just to try to have named arguments, or something else?

Reply:
Yes, it's just to mimic named arguments. I'm not sure which is best, but the rationale was that two consecutive positional string arguments might be more error prone.